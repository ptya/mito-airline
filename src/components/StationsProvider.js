import React, { createContext, useState, useEffect } from "react";
import PropTypes from "prop-types";

const StationsContext = createContext();

const originSession = "mito-origin";
const destSession = "mito-destination";

const StationsProvider = props => {
  const [stations, setStations] = useState();
  const [origin, setOrigin] = useState(
    JSON.parse(sessionStorage.getItem(originSession)) || null
  );
  const [destination, setDestination] = useState(
    JSON.parse(sessionStorage.getItem(destSession)) || null
  );
  const [departureDate, setDepartureDate] = useState();
  const [returnDate, setReturnDate] = useState();
  const [secondaryReturnDate, setSecondaryReturnDate] = useState();

  const fetchStations = async () => {
    const res = await fetch("https://mock-air.herokuapp.com/asset/stations");
    const data = await res.json();
    setStations(data);
  };

  useEffect(() => {
    fetchStations();
  }, []); // empty array to run only on mount

  return (
    <StationsContext.Provider
      value={{
        stations,
        origin,
        destination,
        departureDate,
        returnDate,
        secondaryReturnDate,
        setOrigin,
        setDestination,
        setDepartureDate,
        setReturnDate,
        setSecondaryReturnDate
      }}
    >
      {stations ? props.children : null}
    </StationsContext.Provider>
  );
};

StationsProvider.propTypes = {
  children: PropTypes.node.isRequired
};

export default StationsProvider;
export { StationsContext, originSession, destSession };
