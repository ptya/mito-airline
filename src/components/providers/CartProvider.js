import React, { createContext, useReducer } from "react";
import PropTypes from "prop-types";

const CartContext = createContext();

const initialCart = {
  inbound: null,
  inboundPrice: 0,
  inboundFare: "",
  outbound: null,
  outboundPrice: 0,
  outboundFare: "",
  total: 0
};

const cartReducer = (state, action) => {
  switch (action.type) {
    case "setInbound":
      return {
        ...state,
        inbound: action.flight,
        inboundPrice: action.price,
        inboundFare: action.fare,
        total: state.outboundPrice + action.price
      };
    case "setOutbound":
      return {
        ...state,
        outbound: action.flight,
        outboundPrice: action.price,
        outboundFare: action.fare,
        total: state.inboundPrice + action.price
      };
    case "purge":
      return initialCart;
    default:
      return state;
  }
};

const CartProvider = props => {
  const [cart, cartDispatch] = useReducer(cartReducer, initialCart);

  return (
    <CartContext.Provider
      value={{
        cart,
        cartDispatch
      }}
    >
      {props.children}
    </CartContext.Provider>
  );
};

CartProvider.propTypes = {
  children: PropTypes.node.isRequired
};

export default CartProvider;
export { CartContext, cartReducer, initialCart };
