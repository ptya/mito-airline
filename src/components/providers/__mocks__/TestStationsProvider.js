import React, { useReducer } from "react";
import PropTypes from "prop-types";

import {
  StationsContext,
  stationsReducer,
  initialStations
} from "../StationsProvider";

const TestStationsProvider = props => {
  const { state, reducer } = props;
  const [stations, stationsDispatch] = useReducer(reducer, state);

  return (
    <StationsContext.Provider
      value={{
        stations,
        stationsDispatch
      }}
    >
      {props.children}
    </StationsContext.Provider>
  );
};

TestStationsProvider.propTypes = {
  children: PropTypes.node.isRequired,
  state: PropTypes.object,
  reducer: PropTypes.func
};

TestStationsProvider.defaultProps = {
  state: initialStations,
  reducer: stationsReducer
};

export default TestStationsProvider;
