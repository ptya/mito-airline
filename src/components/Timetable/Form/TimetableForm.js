import React, { useContext } from "react";

// components
import CustomDatePicker from "components/CustomDatePicker";

// context
import { StationsContext } from "components/StationsProvider";

// global elements
import CalendarIco from "components/elements/CalendarIco";
import InputField from "components/elements/InputField";

// local elements
import Form from "./elements/Form";
import SearchButton from "./elements/SearchButton";

const TimetableForm = () => {
  const {
    stationsDispatch,
    stations: { secondaryReturnDate }
  } = useContext(StationsContext);

  const onSubmit = event => {
    event.preventDefault();
    if (secondaryReturnDate) {
      stationsDispatch({ type: "setReturn", returnDate: secondaryReturnDate });
    }
  };

  return (
    <Form autoComplete="off" onSubmit={onSubmit}>
      <fieldset>
        <InputField>
          <CustomDatePicker placeholder="Return" type="secondaryReturn" />
          <CalendarIco />
        </InputField>
        <SearchButton type="submit">Search</SearchButton>
      </fieldset>
    </Form>
  );
};

export default TimetableForm;
