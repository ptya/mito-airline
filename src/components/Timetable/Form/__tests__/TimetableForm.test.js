import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";

import TimetableForm from "../TimetableForm";
import { stationsReducer } from "components/providers/StationsProvider";
import TestStationsProvider from "components/providers/__mocks__/TestStationsProvider";

afterEach(() => {
  cleanup();
  console.error.mockClear();
  mockStationsReducer.mockClear();
});

console.error = jest.fn();
const mockStationsReducer = jest.fn(stationsReducer);

test("<TimetableForm /> renders without error and matches snapshot", () => {
  const { container } = render(
    <TestStationsProvider>
      <TimetableForm />
    </TestStationsProvider>
  );

  expect(console.error).not.toHaveBeenCalled();
  expect(container).toMatchSnapshot();
});

test("Form is properly validated when empty", async () => {
  const { getByTestId, queryByTestId, getByLabelText } = render(
    <TestStationsProvider reducer={mockStationsReducer}>
      <TimetableForm />
    </TestStationsProvider>
  );

  // wait for fetch to finish
  const form = getByTestId("tf-form");
  const input = getByLabelText("Return");

  // there is no error shown yet
  expect(queryByTestId("error")).toBeFalsy();

  // there should be error shown
  fireEvent.submit(form);
  expect(getByTestId("error")).toBeTruthy();

  // fill in the input so error disappears
  fireEvent.change(input, {
    target: {
      value: "2019.07.07"
    }
  });
  expect(mockStationsReducer).toHaveBeenCalledWith(
    expect.any(Object),
    expect.objectContaining({
      type: "setSecondaryReturn"
    })
  );
  expect(queryByTestId("error")).toBeFalsy();

  // delete the input so error reappears
  fireEvent.change(input, {
    target: {
      value: ""
    }
  });
  expect(mockStationsReducer).toHaveBeenLastCalledWith(expect.any(Object), {
    secondaryReturnDate: null,
    type: "setSecondaryReturn"
  });
  expect(getByTestId("error")).toBeTruthy();
});

test("Form can be submitted", async () => {
  const { getByTestId, queryByTestId, getByLabelText } = render(
    <TestStationsProvider reducer={mockStationsReducer}>
      <TimetableForm />
    </TestStationsProvider>
  );

  const form = getByTestId("tf-form");
  const input = getByLabelText("Return");

  // set value
  fireEvent.change(input, {
    target: {
      value: "2019.07.07"
    }
  });

  expect(mockStationsReducer).toHaveBeenCalledWith(
    expect.any(Object),
    expect.objectContaining({
      type: "setSecondaryReturn"
    })
  );
  expect(input.value).toBe("Sun 7. Jul. 2019");

  fireEvent.submit(form);
  expect(mockStationsReducer).toHaveBeenLastCalledWith(
    expect.any(Object),
    expect.objectContaining({
      type: "setReturn"
    })
  );
  expect(queryByTestId("error")).toBeFalsy();
  expect(console.error).not.toHaveBeenCalled();
});
