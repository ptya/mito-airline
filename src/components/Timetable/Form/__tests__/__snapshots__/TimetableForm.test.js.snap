// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TimetableForm /> renders without error and matches snapshot 1`] = `
<div>
  <form
    autocomplete="off"
    class="sc-gzVnrw jDaHzc"
    data-testid="tf-form"
  >
    <fieldset>
      <div
        class="sc-bxivhb sc-ifAKCX cDeBvR"
      >
        <div
          class="sc-htpNat hlaEpc"
          placeholder="Return"
        >
          <label
            for="secondaryReturn"
          >
            Return
          </label>
          <div
            class="react-datepicker-wrapper"
          >
            <div
              class="react-datepicker__input-container"
            >
              <input
                class="datepicker"
                id="secondaryReturn"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <i
          class="sc-bZQynM kulpyt"
        />
      </div>
      <button
        class="sc-htoDjs sc-dnqmqq cgnJCp"
        type="submit"
      >
        Search
      </button>
    </fieldset>
  </form>
</div>
`;
